### DISCUSSION TITLE
Authentication Support for RDF ElasticSearch

### DISCUSSION THREAD

**Discussion by RyanMoss96** [eclipse-rdf4j/rdf4j#4395] [Q&A]
Hi, 

I've been trying to setup RDF4J on a small test project that uses Elasticsearch as its database but seem to be hitting the following error:

`[main] INFO org.eclipse.rdf4j.sail.elasticsearchstore.ElasticsearchStore - Waiting for Elasticsearch to start
[main] INFO org.eclipse.rdf4j.sail.elasticsearchstore.ElasticsearchStore - Unable to connect to elasticsearch cluster due to NoSuchFieldError
java.lang.NoSuchFieldError: LUCENE_8_6_0`

The only thing I can think of that is causing this is that the ElasticSearch instance has a username and password/TLS enabled on it preventing my project from connecting but there doesn't seem to be much documentation available on how to use the ElasticSearch Datastore other than the following line:

`Repository db = new SailRepository(new ElasticsearchStore("localhost", 9200, "elasticsearch", "rdf4j_index"));`

Does anybody know how to authenticate my connection?

**Comment by barthanssens** [comment]
Hmz, would it be possible to add some more info on the RDF4J version you are using and the version of Elasticsearch ?
Lucene 8.6 seems to be somewhat too new, since RDF4J is currently using ES 7.8 (released with Lucene 8.5)

**Reply by RyanMoss96** [reply]
Apologies I got the error wrong and resolved that part. It's defo an auth issue for my next one though, I receive the following output

`[main] INFO org.eclipse.rdf4j.sail.elasticsearchstore.ElasticsearchStore - Waiting for Elasticsearch to start
ERROR StatusLogger Log4j2 could not find a logging implementation. Please add log4j-core to the classpath. Using SimpleLogger to log to the console...
[main] INFO org.eclipse.rdf4j.sail.elasticsearchstore.ElasticsearchStore - Unable to connect to elasticsearch cluster due to NoNodeAvailableException
NoNodeAvailableException[None of the configured nodes are available: [{#transport#-1}{MHj0Q7_4TP2fqTbLga0Auw}{localhost}{127.0.0.1:9200}]]`

My elastic instance outputs the following:

`{"@timestamp":"2023-01-31T23:33:59.073Z", "log.level": "WARN", "message":"received plaintext http traffic on an https channel, closing connection Netty4HttpChannel{localAddress=/172.19.0.2:9200, remoteAddress=/172.19.0.1:56682}", "ecs.version": "1.2.0","service.name":"ES_ECS","event.dataset":"elasticsearch.server","process.thread.name":"elasticsearch[586c9fff871b][transport_worker][T#3]","log.logger":"org.elasticsearch.xpack.security.transport.netty4.SecurityNetty4HttpServerTransport","elasticsearch.cluster.uuid":"YP-m9xgfRqinfoMCG-8cuQ","elasticsearch.node.id":"l7HjKoslSjOtkd7Asfe6bw","elasticsearch.node.name":"586c9fff871b","elasticsearch.cluster.name":"docker-cluster"}`

If I specify https://localhost in my new ElasticsearchStore I get the following output from the Java project:

`[main] INFO org.eclipse.rdf4j.sail.elasticsearchstore.ElasticsearchStore - Waiting for Elasticsearch to start
ERROR StatusLogger Log4j2 could not find a logging implementation. Please add log4j-core to the classpath. Using SimpleLogger to log to the console...
[main] INFO org.eclipse.rdf4j.sail.elasticsearchstore.ElasticsearchStore - Unable to connect to elasticsearch cluster due to SailException
org.eclipse.rdf4j.sail.SailException: java.net.UnknownHostException: https://127.0.0.1: nodename nor servname provided, or not known`

My POM is as followed:

`<dependencies>
		<dependency>
			<groupId>org.eclipse.rdf4j</groupId>
			<artifactId>rdf4j-storage</artifactId>
			<type>pom</type>
			<version>4.2.2</version>
		</dependency>
		<dependency>
			<groupId>org.eclipse.rdf4j</groupId>
			<artifactId>rdf4j-sail-elasticsearch-store</artifactId>
			<version>4.2.2</version>
		</dependency>
		<dependency>
			<groupId>org.slf4j</groupId>
			<artifactId>slf4j-simple</artifactId>
			<version>1.7.36</version>
			<scope>runtime</scope>
		</dependency>
		<dependency>
			<groupId>co.elastic.clients</groupId>
			<artifactId>elasticsearch-java</artifactId>
			<version>8.6.1</version>
		</dependency>
		<dependency>
			<groupId>org.elasticsearch.client</groupId>
			<artifactId>transport</artifactId>
			<version>7.17.8</version>
		</dependency>
		<dependency>
			<groupId>org.apache.lucene</groupId>
			<artifactId>lucene-core</artifactId>
			<version>8.11.2</version>
		</dependency>

		<dependency>
			<groupId>com.fasterxml.jackson.core</groupId>
			<artifactId>jackson-databind</artifactId>
			<version>2.12.3</version>
		</dependency>
	</dependencies>
`

Elastic version is 8.6

**Reply by barthanssens** [reply]
Might be something we've not yet implemented / tested.

RDF4J [SingletonClientProvider](https://github.com/eclipse/rdf4j/blob/main/core/sail/elasticsearch-store/src/main/java/org/eclipse/rdf4j/sail/elasticsearchstore/SingletonClientProvider.java) uses the (now deprecated) ES `PreBuiltTransportClient`  and using HTTPS may require the ES "x-pack"...

I'm _guessing_ using the `PreBuiltXPackTransportClient` and setting `xpack.security.transport.ssl.enabled` might do the trick...
